File: android/FullFeatureSet/TapResearchJavaFullDemo/app/src/main/java/com/tapresearch/tapresearchjavademo/MainActivity.java
package com.tapresearch.tapresearchjavademo;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;

import com.tapresearch.tapsdk.TapInitOptions;
import com.tapresearch.tapsdk.TapResearch;
import com.tapresearch.tapsdk.callback.TRContentCallback;
import com.tapresearch.tapsdk.models.TRError;
import com.tapresearch.tapsdk.models.TRPlacementDetails;
import com.tapresearch.tapsdk.models.TRReward;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;

import kotlinx.serialization.InternalSerializationApi;

@InternalSerializationApi
public class MainActivity extends AppCompatActivity {
    // Tag used for logging. Helps in identifying the source of log messages.
    private static final String LOG_TAG = "MainJavaDemo";
    private TextView statusView;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // User identifier for demo purposes. Replace with actual user ID in production.
        String myUserIdentifier = "tr-sdk-test-user-46183135";
        // API Token retrieved from resources. Replace with actual token.
        String myApiToken = getString(R.string.api_token);

        // Logging API Token and User Identifier for debug purposes.
        Log.d(LOG_TAG, "API Token: " + myApiToken);
        Log.d(LOG_TAG, "User identifier: " + myUserIdentifier);

        HashMap<String, Object> userAttributes = new HashMap<>();
        // Example user attributes. Replace or extend with relevant user data.
        userAttributes.put("age", 25);
        userAttributes.put("VIP", true);
        userAttributes.put("name", "John Doe");
        userAttributes.put("first_seen", Instant.now().toString());

        // Setting the main content view layout.
        setContentView(R.layout.activity_main);
        edgeToEdge(findViewById(R.id.root));
        statusView = findViewById(R.id.statusView);
        ListView listView = findViewById(R.id.listView);
        // Placeholder data until SDK is ready.
        String[] placements = {"Waiting for SDK to be ready..."};
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, placements);
        listView.setAdapter(adapter);

        // Initialize the SDK with necessary parameters including callbacks for rewards and errors.
        TapResearch.INSTANCE.initialize(
                myApiToken, // Replace with your API Token
                myUserIdentifier, // Replace with your User Identifier
                MainActivity.this, // Current activity context
                this::showReward, // Callback to handle reward display
                this::showError, // Callback to handle errors
                // Callback indicating SDK is ready. Used to send user attributes and log readiness.
                () -> {
                    Log.d(LOG_TAG, "SDK Ready");
                    // Sending custom user attributes to the SDK.
                    TapResearch.INSTANCE.sendUserAttributes(userAttributes,
                            false,
                            trError -> statusView.setText("Error sending user attributes: " + trError.toString()));
                    // Setting up the main content view after initialization.
                    runOnUiThread(this::doSetContent);
                },
                // Callback for handling quick questions data.
                (qqData) -> {
                    statusView.setText("QQ Data: " + qqData);
                },
                new TapInitOptions(
                        userAttributes,
                        true // Optional parameter to clear previous user attributes
                ) // Optional initialization options
        );
    }

    // Method to update the content view with actual data after SDK initialization.
    @SuppressLint("SetTextI18n")
    private void doSetContent() {
        statusView.setText("SDK is ready");

        // Retrieve placement tags from resources.
        String[] placements = getResources().getStringArray(R.array.placement_tags);

        ListView listView = findViewById(R.id.listView);

        // Setting up the adapter with placement tags.
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, placements);
        listView.setAdapter(adapter);

        // Handling click events on the list items.
        listView.setOnItemClickListener((parent, view, position, id) -> {
            String selectedItem = placements[position];
            // Checking if content is available for the selected placement before showing.
            if (TapResearch.INSTANCE.canShowContentForPlacement(selectedItem, trError -> Log.d(LOG_TAG, "Whoops " + trError.toString()))) {
                // Custom parameters for the content display. Adjust as needed.
                HashMap<String, Object> customParameters = new HashMap<>();
                customParameters.put("age", 25);
                customParameters.put("VIP", "true");
                customParameters.put("name", "John Doe");
                customParameters.put("first_seen", Instant.now().toString());
                // Showing content for the selected placement with custom parameters.
                TapResearch.INSTANCE.showContentForPlacement(
                        selectedItem,
                        new TRContentCallback() {
                            @Override
                            public void onTapResearchContentShown(String placementTag) {
                                statusView.setText("Content shown");
                            }

                            @Override
                            public void onTapResearchContentDismissed(String placementTag) {
                                statusView.setText("Content dismissed");
                            }
                        },
                        customParameters,
                        this::showError,
                        null
                );
            }
            // Feedback toast for item click.
            statusView.setText("Clicked: " + selectedItem);
        });
    }

    // Method to display an error toast with the provided error description.
    private void showError(TRError error) {
        Log.d(LOG_TAG, "SDK error: " + error.getDescription());
        statusView.setText(error.getDescription());
    }

    // Method to display a reward toast summarizing the rewards earned.
    private void showReward(List<TRReward> rewards) {
        int rewardCount = 0;
        String currencyName = "";
        String eventType = "";
        // Sum up the rewards.
        for (TRReward reward : rewards) {
            Log.d(LOG_TAG, "reward: " + reward);
            if (reward.getRewardAmount() != null) {
                rewardCount += reward.getRewardAmount();
            }
            currencyName += reward.getCurrencyName()+" ";
            eventType += reward.getPayoutEventType()+" ";
        }

        // Displaying the total rewards earned in a toast message.

        statusView.setText("Congrats! You've earned [" + rewardCount + "] [" + currencyName + "]. Event type is " + eventType);
    }

    private void edgeToEdge(View view) {
        ViewCompat.setOnApplyWindowInsetsListener(view, (v, insets) -> {
            Insets bars = insets.getInsets(
                    WindowInsetsCompat.Type.systemBars()
                            | WindowInsetsCompat.Type.displayCutout()
            );
            v.setPadding(bars.left, bars.top, bars.right, bars.bottom);
            return WindowInsetsCompat.CONSUMED;
        });
    }

}

File: android/FullFeatureSet/TapResearchKotlinFullDemo/app/src/main/java/com/tapresearch/tapresearchkotlindemo/MainActivity.kt
package com.tapresearch.tapresearchkotlindemo

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.ui.Modifier
import com.tapresearch.tapresearchkotlindemo.preview.SurveyWallPreviewActivity
import com.tapresearch.tapresearchkotlindemo.ui.MainUi
import com.tapresearch.tapresearchkotlindemo.ui.theme.TapResearchKotlinDemoTheme
import com.tapresearch.tapsdk.TapInitOptions
import com.tapresearch.tapsdk.TapResearch
import com.tapresearch.tapsdk.callback.TRContentCallback
import com.tapresearch.tapsdk.callback.TRErrorCallback
import com.tapresearch.tapsdk.callback.TRQQDataCallback
import com.tapresearch.tapsdk.callback.TRRewardCallback
import com.tapresearch.tapsdk.models.QQPayload
import com.tapresearch.tapsdk.models.TRError
import com.tapresearch.tapsdk.models.TRReward
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.update

class MainActivity : ComponentActivity(), TRRewardCallback {
    val LOG_TAG = "TapKotlinDemo"

    private val initializingStateFlow = MutableStateFlow(true)

    override fun onTapResearchDidReceiveRewards(rewards: MutableList<TRReward>) {
        showRewardToast(rewards)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val myUserIdentifier = "tr-sdk-test-user-46183135" // Try a different user identifier by changing the number at the end
        val myApiToken = "fb28e5e0572876db0790ecaf6c588598" // Insert your API token here

        TapResearch.initialize(
            apiToken = myApiToken,
            userIdentifier = myUserIdentifier,
            activity = this@MainActivity,
            errorCallback = { error -> showErrorToast(error) },
            sdkReadyCallback = {

                initializingStateFlow.update { false }
                Log.d(LOG_TAG, "SDK is ready")
                // you can query placement details when sdk is ready
                // printPlacementDetails()
            },
            rewardCallback = this@MainActivity,
            initOptions = TapInitOptions(
                // Uncomment the following lines to set user attributes
//                userAttributes = hashMapOf(
//                    "is_vip" to 1,
//                ),
//                clearPreviousAttributes = true,
            ),
            qqDataCallback = object : TRQQDataCallback {
                override fun onQuickQuestionDataReceived(data: QQPayload) {
                    Log.d(LOG_TAG, "QQ data received: $data")
                }
            },
        )

        val buttonOptions = resources.getStringArray(R.array.placement_tags)

        setContent {
            TapResearchKotlinDemoTheme {
                // A surface container using the 'background' color from the theme
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background,
                ) {
                    MainUi(
                        userIdentifier = myUserIdentifier,
                        buttonOptions = buttonOptions.toList(),
                        openPlacement = { placementTag ->
                            // Test for Custom Parameters
                            val customParams: HashMap<String, Any> =
                                hashMapOf(
                                    "testKey" to "testValue",
                                )
                            TapResearch.showContentForPlacement(
                                tag = placementTag,
                                customParameters = null,
                                contentCallback =
                                object : TRContentCallback {
                                    override fun onTapResearchContentShown(placementTag: String) {
                                        Log.d(LOG_TAG, "Content shown for $placementTag")
                                    }

                                    override fun onTapResearchContentDismissed(placementTag: String) {
                                        Log.d(
                                            LOG_TAG,
                                            "Content dismissed for $placementTag",
                                        )
                                    }
                                },
                                errorCallback = TRErrorCallback { error -> showErrorToast(error) },
                            )
                        },
                        onSetUserIdentifier = { userId ->
                            TapResearch.setUserIdentifier(
                                userIdentifier = userId,
                            )
                            Toast.makeText(this@MainActivity, "SetUserIdentifier: $userId", Toast.LENGTH_SHORT).show()
                        },
                        sendUserAttributes = {
                            TapResearch.sendUserAttributes(
                                userAttributes =
                                hashMapOf(
                                    "testStatus" to "VIP",
                                    "testInt" to 2,
                                ),
                                errorCallback =
                                    TRErrorCallback { error -> showErrorToast(error) },
                            )
                            Toast.makeText(this@MainActivity, "Sent User Attributes", Toast.LENGTH_SHORT).show()
                            Log.d(LOG_TAG, "Sent User Attributes")
                        },
                        showWallPreview = {
                            startActivity(Intent(this, SurveyWallPreviewActivity::class.java))
                        },
                        onGetPlacementDetailsClicked = {
                            printPlacementDetails()
                        },
                        initializingStateFlow = initializingStateFlow,
                    )
                }
            }
        }
    }

    private fun printPlacementDetails() {
        val placementDetails = TapResearch.getPlacementDetails(
            placementTag = "earn-center",
        ){}
        Log.d(LOG_TAG, "Placement Details: $placementDetails")
        alertDialog("Placement Details: $placementDetails")

    }

    override fun onResume() {
        super.onResume()
        TapResearch.setRewardCallback(this@MainActivity)
    }

    override fun onPause() {
        super.onPause()
        TapResearch.setRewardCallback(null)
    }

    private fun contentDismissed(tag: String) {
        Log.d(LOG_TAG, "dismissed: $tag")
    }

    private fun contentShown(tag: String) {
        Log.d(LOG_TAG, "shown: $tag")
    }

    private fun showErrorToast(error: TRError) {
        initializingStateFlow.update{false}
        Log.d(LOG_TAG, "error: $error")
        alertDialog(error.description?:"Unknown Error")
    }

    private fun showRewardToast(rewards: MutableList<TRReward>) {
        Log.d(LOG_TAG, "rewards: $rewards")
        var rewardAmount = 0f
        for (reward: TRReward in rewards) {
            Log.d(LOG_TAG, "reward: $reward")
            Log.d(LOG_TAG, "Amount: ${reward.rewardAmount}")
            reward.rewardAmount?.let { rewardAmount += it }
        }

        val currencyName = rewards.first().currencyName
        val eventType = rewards.first().payoutEventType
        alertDialog("Congrats! You've earned $rewardAmount $currencyName. Event type is $eventType")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(LOG_TAG, "onDestroy()")
    }

    private fun alertDialog(message: String) {
        val builder = AlertDialog.Builder(this)
            .setMessage(message)
            .setPositiveButton("OK") { dialog, _ ->
                dialog.dismiss()
            }
        val alertDialog = builder.create()
        alertDialog.show()
    }

}
